### ISSUE TITLE
Disabled CA certificates on android are trusted by Dart

### ISSUE THREAD

**Issue by brianquinlan** [dart-lang/sdk#48056]
On Android OS, the list of Root Certificates loaded by the DartSDK SecurityContext is loaded from the directory /system/etc/security/cacerts.

The Root Certificate files placed in this path will match the System Certificates in the Trusted credentials on the Android OS. 

However, even if Certificate is set to Disable on the Android OS, the Root Certificate file will not be removed from /system/etc/security/cacerts. So even if a user sets any Root Certificate to be untrusted, the DartSDK will ignore that setting and trust it.

Steps to reproduce:

1. On AndroidOS, disable the System Root Certificate DigiCert High Assurance EV Root CA in Trusted credentials.
2. Using the `http` package of DartSDK, connect to the server where DigiCert High Assurance EV Root CA will be the RootCA. An example of the code is shown below.
```
dart
// RootCA for www.intellilink.co.jp FQDN is DigiCert High Assurance EV Root CA
http.get(Uri.parse('https://www.intellilink.co.jp'));
```
3. For the reasons explained above, DartSDK trusts the Root Certificate with the disabled setting and establishes communication with it. Note that NET::ERR_CERT_AUTHORITY_INVALID is generated when you connect to https://www.intellilink.co.jp with Google Chrome, which confirms that the above disabled setting is working.

**Comment by acoutts** [comment]
Could this also mean if a user has explicitly added a new certificate to their system, dart wouldn't trust it?
One issue I'm seeing in the wild is users on corporate public wifi networks are getting certificate errors in my flutter app but they are clearly able to use the internet otherwise to contact our support team, which means their device must trust the root CA being used but the flutter app is not.

**Comment by komaxx** [comment]
@acoutts
Yes, many corporate Wifi network setups will not work with Dart/Flutter (at least with Android devices).
In one app, we've built a workaround where a required certificate would be installed into the app per MDM additionally to installing it in the OS itself.
That is brittle workaround, though. 

We've also tried to pipe all traffic not through `dart.http` but wrote native plugin code for that instead. This will fail for everything that you don't control directly, though, like image loading or video streaming.

In the end, we scrapped the Flutter project for all business apps, and started re-implementing natively.

**Comment by brianquinlan** [comment]
Hey @komaxx,

I'd like to understand your use case (and I'm not familiar with how typical corporate networks do security).

So the corporate network that you are trying to access requires a client certificate. When you were explicitly setting the certificate in your app, were you using [`SecurityContext.useCertificateChainBytes`](https://api.dart.dev/stable/2.18.4/dart-io/SecurityContext/useCertificateChainBytes.html)?

Would [`getCertificateChain`](https://developer.android.com/reference/android/security/KeyChain#getCertificateChain(android.content.Context,%20java.lang.String) have provided the certificate that you needed?

What did your native plugin do? Provide the needed certificate, do the entire HTTP transport or something else?

Thanks for your help!

**Comment by aam** [comment]
> Could this also mean if a user has explicitly added a new certificate to their system, dart wouldn't trust it?

It depends on how and kind of a certificate was added to the system.
Generally speaking, by default dart app will pick up _trusted_root_certs_ from the system, so if a user adds new new trusted root cert to the system, dart app will use it. Dart developer can also allows force use of trusted roots certs that are built-in into the dart vm(via [`SecurityContext(withTrustedRoots:true)`](https://api.flutter.dev/flutter/dart-io/SecurityContext/SecurityContext.html)). Dart developer can also explicitly specify certificates that should be used(via [`SecurityContext.setTrustedCertificates`](https://api.flutter.dev/flutter/dart-io/SecurityContext/setTrustedCertificates.html)

> One issue I'm seeing in the wild is users on corporate public wifi networks are getting certificate errors in my flutter app but they are clearly able to use the internet otherwise to contact our support team, which means their device must trust the root CA being used but the flutter app is not.

I recommend opening new issue for the issue you see, since this one is specifically about handling of disabled certs on Android, and providing more details along the lines @brianquinlan suggested.

**Comment by brianquinlan** [comment]
Hey @aam,

Maybe I'm misunderstanding but Dart is loading certs from [`/system/etc/security/cacerts`](https://github.com/dart-lang/sdk/blob/main/runtime/bin/security_context_android.cc#L50).

I think that user-add certificates go in `/data/misc/keystore/user_0/` (or in a database in new Android versions), which we don't consider.

Good point about the new issue - if this is looking like a client certificate issue then I'll file a new issue just about that.

**Comment by aam** [comment]
> I think that user-add certificates go in `/data/misc/keystore/user_0/` (or in a database in new Android versions), 

Ah, I see. If that's the case, then you are right, we are not going to pick those certificates up automatically at the moment

**Comment by komaxx** [comment]
Hi @brianquinlan 

Happy to help! 
Here's what I've encountered as a common corporate network setup: (Android) devices are controlled via some sort of MDM (Knox, MobileIron, ..). Via these MDMs, self-signed certificates are installed on the phones. In the settings app on the phone, these will then show up as user-added certificates.
Also Wifi connections are often set-up to use a man-in-the-middle proxy (for "security") which uses again a self-signed certificate for TLS.

For the **browser**, everything works then out-of-the-box: Browser apps by default trust user-added certificates.

Fot **native Android** apps, there is only a small change necessary: One just has to add a security config XML file that enables user-installed certificates as trust-anchors. Proxy settings are honored automatically.

For **Flutter apps**, we're now in hot water: 
- First, Flutter will not respect the phones proxy settings (see: https://github.com/flutter/flutter/issues/26359) 
- Then, when a workaround for that was made, the user-installed certificates will not be accepted. For this, we could not find a simple solution, as there seems to be no easy way to read user-certificates from the system. 

With one customer we could build a scheme where the PEM of the used self-signed certificate would be transported to the app via MDM. Then, we simply registered an `HttpOverride` and created a `HttpClient` that would check in its `badCertificateCallback` against this pem. It works - but it puts our app at a disadvantage against competitor's apps that don't need fiddling around with certificates in MDM interfaces.

The second workaround was to simply dump all http requesting into a plugin and run the requests natively. That gets old quickly (multipart requets, custom headers, error handling, ..) as we had to re-implement parts of the `http` interface, but it works - for http requests that we do ourself; implicit requests (e.g. `Image.network`) fail. One can write workarounds for all those things as well. However, we decided at that point to go for a different platform for apps targeted at corporate customers.


I hope that makes things a bit clearer?

**Comment by brianquinlan** [comment]
Based on the discussion, I've filed two new bugs:
- [HttpClient should honor the system configured proxy](https://github.com/dart-lang/sdk/issues/50434)
- [HttpClient should honor user-installed certificates on Android](https://github.com/dart-lang/sdk/issues/50435)

**Comment by tomekit** [comment]
We're facing similar issue where Dart/Flutter doesn't respect configured user managed trusted CA. It affects at least Android and Linux platforms.

In more detail, we're getting: `HandshakeException: Handshake error in client (OS Error: \n\tCERTIFICATE_VERIFY_FAILED: unable to get local issuer certificate(handshake.cc:393`

On Ubuntu certs are added like this:
```
sudo cp rootCA.crt /usr/local/share/ca-certificates
sudo update-ca-certificates
```
and `curl https://mydomain.com` command confirms that certificate is installed successfully.

On Android we've installed CA and tested on on Android 11 and Android 13 without success:

![Screenshot from 2023-06-02 11-50-25](https://github.com/dart-lang/sdk/assets/7759113/75e9894d-4d36-4def-8b49-ee16ac17e559)

![Screenshot from 2023-06-02 11-50-49](https://github.com/dart-lang/sdk/assets/7759113/d7c624d9-4cd9-406a-9152-066e5c2ef522)

This is an example how we generate the CA and use it to sign a certificate for: `mydomain.com`

```
openssl genrsa -des3 -out rootCA.key 4096
openssl req -x509 -new -nodes -key rootCA.key -sha256 -days 1024 -out rootCA.crt

openssl genrsa -out mydomain.com.key 2048

openssl req -new -sha256 \
    -key mydomain.com.key \
    -subj "/C=US/ST=CA/O=MyOrg, Inc./CN=mydomain.com" \
    -reqexts SAN \
    -config <(cat /etc/ssl/openssl.cnf \
        <(printf "\n[SAN]\nsubjectAltName=DNS:mydomain.com,DNS:www.mydomain.com")) \
    -out mydomain.com.csr

openssl x509 -req -in mydomain.com.csr -CA rootCA.crt -CAkey rootCA.key -CAcreateserial -out mydomain.com.crt -days 500 -sha256
```

Are there any plans to address the issue, so Flutter correctly recognizes the CA installed on the system?

On our side it's user trying to connect to their locally managed endpoints using HTTPS, however I am sure that this is a deal breaker for many corporate apps / use cases, where CA is managed by the IT department and installed on the user's end device.
